src/main/java/com/direwolf20/buildinggadgets/common/entities/BlockBuildEntity.java

```diff
@@ -21,62 +20,105 @@
+            this.id = id;
+        }
+
+        public String getID() {
+            return id;
+        }
+
+        public abstract void onBuilderEntityDespawn(BlockBuildEntity builder);
+    }
+
+    private static final DataParameter<String> MODE = EntityDataManager.createKey(BlockBuildEntity.class, DataSerializers.STRING);
     private static final DataParameter<Optional<IBlockState>> SET_BLOCK = EntityDataManager.createKey(BlockBuildEntity.class, DataSerializers.OPTIONAL_BLOCK_STATE);
-    private static final DataParameter<BlockPos> FIXED = EntityDataManager.createKey(BlockBuildEntity.class, DataSerializers.BLOCK_POS);
-    private static final DataParameter<Boolean> usePaste = EntityDataManager.createKey(BlockBuildEntity.class, DataSerializers.BOOLEAN);
+    private static final DataParameter<Boolean> USE_PASTE = EntityDataManager.createKey(BlockBuildEntity.class, DataSerializers.BOOLEAN);
 
     private IBlockState setBlock;
     private IBlockState originalSetBlock;
-    private IBlockState actualSetBlock;
     private EntityLivingBase spawnedBy;
 
-    private int mode;
+    private Mode mode;
     private boolean useConstructionPaste;
 
     public BlockBuildEntity(World world) {
         super(BGEntities.BUILD_BLOCK, world);
     }
 
-    public BlockBuildEntity(World world, BlockPos spawnPos, EntityLivingBase player, IBlockState spawnBlock, int toolMode, IBlockState actualSpawnBlock, boolean constrPaste) {
+    public BlockBuildEntity(World world, BlockPos spawnPos, EntityLivingBase player, IBlockState spawnBlock, Mode mode, boolean usePaste) {
         this(world);
         setPosition(spawnPos.getX(), spawnPos.getY(), spawnPos.getZ());
 
         IBlockState currentBlock = world.getBlockState(spawnPos);
         TileEntity te = world.getTileEntity(spawnPos);
-
         targetPos = spawnPos;
-        setBlock = te instanceof ConstructionBlockTileEntity ? te.getBlockState() : spawnBlock;
         originalSetBlock = spawnBlock;
 
-        setSetBlock(setBlock);
-
-        if (toolMode == 3) {
+        if (mode == Mode.REPLACE)
             setBlock = te instanceof ConstructionBlockTileEntity ? te.getBlockState() : currentBlock;
-            setSetBlock(setBlock);
-        }
+        else
+            setBlock = te instanceof ConstructionBlockTileEntity ? te.getBlockState() : spawnBlock;
+        setSetBlock(setBlock);
 
-        mode = toolMode;
-        setToolMode(toolMode);
+        this.mode = mode;
+        setToolMode(mode);
 
         spawnedBy = player;
-        actualSetBlock = actualSpawnBlock;
         world.setBlockState(spawnPos, BGBlocks.effectBlock.getDefaultState());
 
-        setUsingConstructionPaste(constrPaste);
+        setUsingPaste(usePaste);
     }
 
     @Override
     protected int getMaxLife() {
         return 20;
     }
 
-    public int getToolMode() {
-        return dataManager.get(toolMode);
+    public Mode getToolMode() {
+        return Mode.valueOf(dataManager.get(MODE));
```

Comment:
This will not work - valueOf looks for the toString value, but you assign your own String id

Reply:
Mark resolved because switching to the number ID method.